source/Mission.cpp

```diff
@@ -1148,57 +1148,60 @@ bool Mission::HasShip(const shared_ptr<Ship> &ship) const
 // about it. This may affect the mission status or display a message.
 void Mission::Do(const ShipEvent &event, PlayerInfo &player, UI *ui)
 {
-	if(event.TargetGovernment()->IsPlayer() && !hasFailed)
+	if(!HasFailed(player))
 	{
-		bool failed = false;
-		string message = "Your ship \"" + event.Target()->Name() + "\" has been ";
-		if(event.Type() & ShipEvent::DESTROY)
+		if(event.TargetGovernment()->IsPlayer())
 		{
-			// Destroyed ships carrying mission cargo result in failed missions.
-			// Mission cargo may have a quantity of zero (i.e. 0 mass).
-			for(const auto &it : event.Target()->Cargo().MissionCargo())
-				failed |= (it.first == this);
-			// If any mission passengers were present, this mission is failed.
-			for(const auto &it : event.Target()->Cargo().PassengerList())
-				failed |= (it.first == this && it.second);
-			if(failed)
-				message += "lost. ";
-		}
-		else if(event.Type() & ShipEvent::BOARD)
-		{
-			// Fail missions whose cargo is stolen by a boarding vessel.
-			for(const auto &it : event.Actor()->Cargo().MissionCargo())
-				failed |= (it.first == this);
-			if(failed)
-				message += "plundered. ";
-		}
+			bool failed = false;
+			string message = "Your ship \"" + event.Target()->Name() + "\" has been ";
+			if(event.Type() & ShipEvent::DESTROY)
+			{
+				// Destroyed ships carrying mission cargo result in failed missions.
+				// Mission cargo may have a quantity of zero (i.e. 0 mass).
+				for(const auto &it : event.Target()->Cargo().MissionCargo())
+					failed |= (it.first == this);
+				// If any mission passengers were present, this mission is failed.
+				for(const auto &it : event.Target()->Cargo().PassengerList())
+					failed |= (it.first == this && it.second);
+				if(failed)
+					message += "lost. ";
+			}
+			else if(event.Type() & ShipEvent::BOARD)
+			{
+				// Fail missions whose cargo is stolen by a boarding vessel.
+				for(const auto &it : event.Actor()->Cargo().MissionCargo())
+					failed |= (it.first == this);
+				if(failed)
+					message += "plundered. ";
+			}
 
-		if(failed)
-		{
-			hasFailed = true;
-			if(isVisible)
-				Messages::Add(message + "Mission failed: \"" + displayName + "\".", Messages::Importance::Highest);
+			if(failed)
+			{
+				hasFailed = true;
+				if(isVisible)
+					Messages::Add(message + "Mission failed: \"" + displayName + "\".", Messages::Importance::Highest);
+			}
 		}
-	}
 
-	if((event.Type() & ShipEvent::DISABLE) && event.Target().get() == player.Flagship())
-		Do(DISABLED, player, ui);
+		if((event.Type() & ShipEvent::DISABLE) && event.Target() == player.FlagshipPtr())
+			Do(DISABLED, player, ui);
```

Comment:
The flagship always belongs to the player, so you can move this to the `if(event.TargetGovernment()->IsPlayer())` block above. ~~yeah ik it wasn't there before, but since you're changing it anyway...~~

Reply:
I don't really see the point tbh, we aren't using it anywhere else.